cvat-core/src/requests-manager.ts

```diff
@@ -0,0 +1,357 @@
+        const result = await serverProxy.requests.list();
+        const requests = [];
+        for (const request of result) {
+            requests.push(
+                new Request({
+                    ...request,
+                }),
+            );
+        }
+        return { requests, count: requests.length };
+    }
+
+    async listen(
+        id: string | null,
+        options?: {
+            callback?: (request: Request) => void,
+            filter?: RequestsFilter,
+            initialRequest?: Request,
+        },
+    ): Promise<Request> {
+        let storedID = null;
+        const params = options?.filter ? fieldsToSnakeCase(options.filter) : {};
+        if (id) {
+            storedID = id;
+        } else if (options?.filter) {
+            storedID = new URLSearchParams(params).toString();
+        }
+
+        if (!storedID) {
+            throw new Error('Neither request id nor filter is provided');
+        }
+        const callback = options?.callback;
+        const initialRequest = options?.initialRequest;
+
+        if (storedID in this.listening) {
+            if (callback) {
+                this.listening[storedID].onUpdate.push(callback);
+            }
+            return this.listening[storedID].promise;
+        }
+        const promise = new Promise<Request>((resolve, reject) => {
+            const timeoutCallback = (): void => {
+                if (!(storedID in this.listening)) {
+                    return;
+                }
+
+                // We make sure that no more than REQUESTS_COUNT requests are sent simultaneously
+                // If thats the case, we re-schedule the timeout
+                const timestamp = Date.now();
+                if (this.requestStack.length >= REQUESTS_COUNT) {
+                    const timestampToCheck = this.requestStack[this.requestStack.length - 1];
+                    const delay = this.delayFor(storedID);
+                    if (timestamp - timestampToCheck < delay) {
+                        this.listening[storedID].timeout = window.setTimeout(timeoutCallback, delay);
+                        return;
+                    }
+                }
+                if (this.requestStack.length >= REQUESTS_COUNT) {
+                    this.requestStack.pop();
+                }
+                this.requestStack.unshift(timestamp);
+
+                serverProxy.requests.status(id, params).then((serializedRequest) => {
+                    const request = new Request({ ...serializedRequest });
+                    const { status } = request;
+                    // check it was not cancelled
+                    if (storedID in this.listening) {
+                        const { onUpdate } = this.listening[storedID];
+                        if ([RQStatus.QUEUED, RQStatus.STARTED, RQStatus.DEFERRED].includes(status)) {
+                            onUpdate.forEach((update) => update(request));
```

Comment:
We can't guarantee one of update callback do not change the request object.

In this case, I will suggest different ways to ensure that it does not happen:
1. Do Request object fully read only
2. Create new Request instance each time for each callback
3. Using typescript define that this argument passes as read only

Reply:
Made the Request read-only